src/dataframe/dataframe.jl

```diff
@@ -1508,3 +1508,63 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+
+function expanddf(df::AbstractDataFrame, indexcols::Array{Symbol,1})
```

Comment:
I would leave `indexcols` either undefined or a wide definition accepting all we currently accept: `ColumnIndex`, `AbstractVector{<:ColumnIndex`, `Regex`, `Not`, `Colon`.

Reply:
also we might add an option to skip rows with `missing` values in `indexcols` like in `groupby`

Reply:
An interesting related question is how to support `nesting` as in dplyr, which allows specifying that a combination of some variables should not be expanded for values which don't appear in the data. This can be a future extension, but we need to make sure we will be able to add it with the current interface. Maybe we could just support that by passing tuples of columns inside the array?

Reply:
@nalimilan I would probably add that as an extension in the future, but will definitely keep it in mind. That said, DataFrames doesn't have an equivalent concept of `nest()`, does it? I'm trying to think of what structure would be passed in to indicate nested columns, or whether this PR should support creating the nesting itself.

Reply:
dplyr's `nesting` in `expand` is completely different from tidyr's `nest` AFAICT.